llvm/lib/Transforms/Scalar/IndVarSimplify.cpp

```diff
@@ -1049,9 +1049,28 @@ linearFunctionTestReplace(Loop *L, BasicBlock *ExitingBB,
     if (Extended) {
       bool Discard;
       L->makeLoopInvariant(ExitCnt, Discard);
-    } else
+    } else{
       CmpIndVar = Builder.CreateTrunc(CmpIndVar, ExitCnt->getType(),
                                       "lftr.wideiv");
+
+      // Set the correct wrap flag to avoid the masking issue.
+      Instruction *TruncInst = dyn_cast<Instruction>(CmpIndVar);
+      
+      // The TruncatedIV is incrementing.
+      if (const SCEVAddRecExpr *TruncAR =
+              dyn_cast<SCEVAddRecExpr>(TruncatedIV)) {
```

Comment:
Because it's an induction variable, is `TruncatedIV` always guaranteed to be a SCEVAddRecExpr? Does it work if we do a cast instead?


```suggestion
      auto *TruncAR = cast<SCEVAddRecExpr>(TruncatedIV);
```